model/atmosphere/advection/src/icon4py/model/atmosphere/advection/advection.py

```diff
@@ -0,0 +1,729 @@
+            grid=self.grid,
+            config=self.config,
+            interpolation_state=self.interpolation_state,
+            least_squares_state=self.least_squares_state,
+            metric_state=self.metric_state,
+            edge_params=self.edge_params,
+            cell_params=self.cell_params,
+            exchange=self.exchange,
+        )
+
+        log.debug("advection class init - end")
+
+    def _allocate_temporary_fields(self):
+        #: intermediate density times cell thickness, includes either the horizontal or vertical advective density increment [kg/m^2]
+        self.rhodz_ast2 = field_alloc.allocate_zero_field(dims.CellDim, dims.KDim, grid=self.grid)
+
+    def run(
+        self,
+        diagnostic_state: advection_states.AdvectionDiagnosticState,
+        prep_adv: advection_states.AdvectionPrepAdvState,
+        p_tracer_now: fa.CellKField[wpfloat],
+        p_tracer_new: fa.CellKField[wpfloat],
+        dtime: wpfloat,
+    ):
+        """
+        Run an advection step.
+
+        Args:
+            diagnostic_state: output argument, data class that contains diagnostic variables
+            prep_adv: input argument, data class that contains precalculated advection fields
+            p_tracer_now: input argument, field that contains current tracer mass fraction
+            p_tracer_new: output argument, field that contains new tracer mass fraction
+            dtime: input argument, the time step
+
+        """
+        log.debug("advection run - start")
+
+        log.debug("communication of prep_adv cell field: mass_flx_ic - start")
+        self.exchange.exchange_and_wait(dims.CellDim, prep_adv.mass_flx_ic)
+        log.debug("communication of prep_adv cell field: mass_flx_ic - end")
+
+        if (
+            self.config.horizontal_advection_type == HorizontalAdvectionType.NO_ADVECTION
+            and self.config.vertical_advection_type == VerticalAdvectionType.NO_ADVECTION
+        ):
+            log.debug("running stencil copy_cell_kdim_field - start")
+            copy_cell_kdim_field.copy_cell_kdim_field(
+                p_tracer_now,
+                p_tracer_new,
+                horizontal_start=self.start_cell_nudging,
+                horizontal_end=self.end_cell_local,
+                vertical_start=0,
+                vertical_end=self.grid.num_levels,
+                offset_provider=self.grid.offset_providers,
+            )
+            log.debug("running stencil copy_cell_kdim_field - end")
+
+            log.debug("advection run - early end")
+            return
+
+        # Godunov splitting
+        if self.even_timestep:  # even timestep, vertical transport precedes horizontal transport
+            # reintegrate density with vertical increment for conservation of mass
+            log.debug("running stencil apply_vertical_density_increment - start")
+            apply_vertical_density_increment.apply_vertical_density_increment(
```

Comment:
you could even fuse/combine the `apply_vertical_density_increment` and `apply_horizontal_density_increment` as they have the same signature by adding the `even_timestep` flag to the signature.  You would then have

```
def run(...)
    self.apply_densiity_increment
    if even_time stemp:
         self._vertical.run(...)
         self._horizontal.run(...)
   else:
        self._horizontal.run()
        self._vertical.run()